sdk/core/Azure.Core.TestFramework/src/Instrumentation/DiagnosticScopeValidatingInterceptor.cs

```diff
@@ -14,94 +14,144 @@ namespace Azure.Core.TestFramework
+                return;
             }
-            else
+
+            invocation.Proceed();
+        }
+
+        internal static void WrapAsyncResult(IInvocation invocation, object target, MethodInfo interceptorMethod)
+        {
+            Type genericArgument = typeof(object);
+            Type awaitableType = invocation.Method.ReturnType;
+            if (invocation.Method.ReturnType is {IsGenericType: true, GenericTypeArguments: {Length: 1} genericTypeArguments})
             {
-                invocation.Proceed();
+                genericArgument = genericTypeArguments[0];
+                awaitableType = invocation.Method.ReturnType.GetGenericTypeDefinition();
             }
+
+            WrapAsyncResultCoreMethod.MakeGenericMethod(genericArgument)
+                .Invoke(null, new object[]
+                {
+                    invocation,
+                    awaitableType,
+                    interceptorMethod
+                        .MakeGenericMethod(genericArgument)
+                        .CreateDelegate(typeof(AsyncCallInterceptor<>).MakeGenericType(genericArgument), target)
+                });
         }
 
-        internal static void AwaitAndValidateDiagnosticScope<T>(Type genericType, IInvocation invocation, string expectedName, string methodName, bool strict)
+        internal static void WrapAsyncResultCore<T>(IInvocation invocation, Type genericType, AsyncCallInterceptor<T> wrap)
         {
             // All this ceremony is not to await the returned Task/ValueTask syncronously
             // instead we are replacing the invocation.ReturnValue with the ValidateDiagnosticScope task
             // but we need to make sure the types match
             if (genericType == typeof(Task<>))
             {
-                invocation.ReturnValue = ValidateDiagnosticScope(async () =>
+                async ValueTask<T> Await()
                 {
                     invocation.Proceed();
-                    return (await (Task<T>)invocation.ReturnValue, false);
-                }, expectedName, methodName, strict).AsTask();
+                    return await (Task<T>)invocation.ReturnValue;
+                }
+
+                invocation.ReturnValue = wrap(invocation, Await).AsTask();
             }
             else if (genericType == typeof(Task))
             {
-                invocation.ReturnValue = ValidateDiagnosticScope<object>(async () =>
+                async ValueTask<T> Await()
                 {
                     invocation.Proceed();
                     await (Task)invocation.ReturnValue;
                     return default;
-                }, expectedName, methodName, strict).AsTask();
+                }
+
+                invocation.ReturnValue = wrap(invocation, Await).AsTask();
             }
             else if (genericType == typeof(ValueTask<>))
             {
-                invocation.ReturnValue = ValidateDiagnosticScope(async () =>
+                ValueTask<T> Await()
```

Comment:
Maybe I'm missing something obvious, by why is this one sync and the one below async? `ValueTask` and `ValueTask<T>` are used in cases where they may be completed, but might not be.

Reply:
The code this replaces was also awaited.

Reply:
>Maybe I'm missing something obvious

This code is the very opposite of obvious :)

The goal of these Await methods to covert various awaitable types to `ValueTask<T>` that we can pass to the wrapping function. All other types `Task/ValueTask/Task<T>` need to be awaited to convert them to `ValueTask<T>`  but since in this case we already have `ValueTask<T>` we just return it.

Reply:
Ah, that makes sense now that you explain it. 🙂